File: other/remove_element.cpp
#include "leetcode.h"

class Solution {
public:
  int removeElement(vector<int> &nums, int val) {
    int idx = 0, k = nums.size();
    while (idx < k) {
      if (nums[idx] == val) {
        //          cout << "idx " << idx << endl;
        //          cout << "k " << k << endl;
        nums[idx] = nums[k - 1];
        --k;
      } else
        ++idx;
    }
    return k;
  }
};

int main() {
  Solution obj;
  vector<int> nums = {0, 1, 2, 2, 3, 0, 4, 2};
  int val = 2;
  for (int i = 0; i < nums.size(); i++)
    cout << nums[i] << ' ';
  cout << endl << nums.size() << endl;
  cout << obj.removeElement(nums, val);
}

File: other/remove_element.rs
use std::convert::TryInto;
struct Solution;

impl Solution {
    pub fn remove_element(mut nums: Vec<i32>, val: i32) -> i32 {
        let (mut idx, mut k) = (0, nums.len());
        while idx < k {
            if nums[idx] == val {
                nums[idx] = nums[k - 1];
                k -= 1;
            } else {
                idx += 1;
            }
        }
        k.try_into().unwrap()
    }
}

fn main() {
    let nums = vec![0,1,2,2,3,0,4,2];
    print!("{}", Solution::remove_element(nums, 2));
}
